pyteal/ast/widemath.py

```diff
@@ -157,3 +439,515 @@ def has_return(self):
+            def __init__(self, lhs: WideUint128, arg: WideUint128):
+                WideUint128.__init__(self, 2)
+                self.term = arg
+                self.lhs = lhs
+
+            def __teal__(self, options: "CompileOptions"):
+                lhsSrt, lhsEnd = self.lhs.__teal__(options)
+                mulSrt, mulEnd = multU128(self, self.term, options)
+                lhsEnd.setNextBlock(mulSrt)
+                return lhsSrt, mulEnd
+
+            def __str__(self) -> str:
+                return "(mulW {} {})".format(self.lhs, self.term)
+
+        return WideUint128MulU128(self, other)
+
+    def mulU64(self, other: Expr):
+        require_type(other, TealType.uint64)
+
+        class WideUint128MulU64(WideUint128):
+            def __init__(self, lhs: WideUint128, arg: Expr):
+                WideUint128.__init__(self, 2)
+                self.term = arg
+                self.lhs = lhs
+
+            def __teal__(self, options: "CompileOptions"):
+                lhsSrt, lhsEnd = self.lhs.__teal__(options)
+                mulSrt, mulEnd = multU128U64(self, self.term, options)
+                lhsEnd.setNextBlock(mulSrt)
+                return lhsSrt, mulEnd
+
+            def __str__(self) -> str:
+                return "(mulW {} {})".format(self.lhs, self.term)
+
+        return WideUint128MulU64(self, other)
+
+    def __truediv__(self, other: Expr):
+        if isinstance(other, WideUint128):
+            return self.divU128(other)
+        else:
+            require_type(other, TealType.uint64)
+            return self.divU64(other)
+
+    def divU128(self, other: "WideUint128"):
+        if not isinstance(other, WideUint128):
+            raise TealInputError("expected WideUint128 input for divU128")
+
+        class WideUint128DivU128(WideUint128):
+            def __init__(self, lhs: WideUint128, arg: WideUint128):
+                WideUint128.__init__(self, 2)
+                self.term = arg
+                self.lhs = lhs
+
+            def __teal__(self, options: "CompileOptions"):
+                lhsSrt, lhsEnd = self.lhs.__teal__(options)
+                termSrt, termEnd = self.term.__teal__(options)
+                lhsEnd.setNextBlock(termSrt)
+                divFromDivmodW = TealSimpleBlock(
+                    [
+                        TealOp(self, Op.divmodw),
+                        TealOp(self, Op.pop),
+                        TealOp(self, Op.pop),
+                    ]
+                )
+                termEnd.setNextBlock(divFromDivmodW)
+                return lhsSrt, divFromDivmodW
+
+            def __str__(self) -> str:
+                return "(divW {} {})".format(self.lhs, self.term)
+
+        return WideUint128DivU128(self, other)
+
+    def divU64(self, other: Expr):
+        require_type(other, TealType.uint64)
+
+        class WideUint128DivU64(WideUint128):
+            def __init__(self, lhs: WideUint128, arg: Expr):
+                WideUint128.__init__(self, 1)
```

Comment:
@ahangsu Request for clarification:  In the following scenarios, I _think_ programs generated via PyTeal do _not_ produce the intended on-chain results.  It's quite possible I have gaps in my understanding.  Can you double check me?
* Use `WideUint128DivU64` for subsequent `WideUint128` arithmetic (e.g. `+`).
  * Concern:  The result is 1 uint64.  Subsequent arithmetic operations assume a `WideUint128` produces 2 outputs (2 uint64).
* Produce `WideUint128` with `outputNum=2`, and supply the value as argument to `@Subroutine`.
  * Concern:  I'm concerned only 1 of 2 stack values will be passed into the subroutine.  I didn't fully vet the concern before raising.  Trying to gauge if you've considered it before I go deeper.

After addressing the above concerns, taking a step back:  
* Is the intent to provide a general purpose wide arithmetic API or to _only_ provide the operations needed to use wide arithmetic operations for intermediate steps? 
* It feels to me like the PR attempts to provide a general purpose wide arithmetic API.  Though it's unclear to me that the goal is a general purpose API.